### PR TITLE
Return 404 or member list when getting joined_members after leaving

### PR DESCRIPTION
**Pull request by andrewdoh**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: changelog.d/13374.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fixed bug when user tries getting joined_members after already having left the room. Contributed by @andrewdoh.
```
Comment (babolivier) [inline changelog.d/13374.bugfix:1]:
```suggestion
Fix a bug introduced in Synapse 0.24.0 that would respond with the wrong error status code to `/joined_members` requests when the requester is not a current member of the room. Contributed by @andrewdoh.
```

File: tests/rest/admin/test_room.py [hunk 2]
```diff
@@ -1772,6 +1772,21 @@ def _set_canonical_alias(
...
+
+        # create the room
+        user_1 = self.register_user("foo", "pass")
+        user_tok_1 = self.login("foo", "pass")
+        room_id_1 = self.helper.create_room_as(user_1, tok=user_tok_1)
+        self.helper.leave(room_id_1, user_1, tok=user_tok_1)
+
+        # delete the rooms and get joined roomed membership
+        url = f"/_matrix/client/r0/rooms/{room_id_1}/joined_members"
+        channel = self.make_request("GET", url.encode("ascii"), access_token=user_tok_1)
```
Comment (babolivier) [inline tests/rest/admin/test_room.py:1786]:
```suggestion
        user = self.register_user("foo", "pass")
        user_tok = self.login("foo", "pass")
        room_id = self.helper.create_room_as(user, tok=user_tok)
        self.helper.leave(room_id, user, tok=user_tok)

        # delete the rooms and get joined roomed membership
        url = f"/_matrix/client/r0/rooms/{room_id}/joined_members"
        channel = self.make_request("GET", url.encode("ascii"), access_token=user_tok)
```

There are only one user and room in this test so let's not make it confusing by inferring there are more.

File: tests/rest/admin/test_room.py [hunk 3]
```diff
@@ -1772,6 +1772,21 @@ def _set_canonical_alias(
             tok=admin_user_tok,
         )
 
+    def test_get_joined_members_after_leave_room(self) -> None:
+        """Test that room members can be requested correctly"""
```
Comment (babolivier) [inline tests/rest/admin/test_room.py:1776]:
```suggestion
        """Test that requesting room members after leaving the room raises a 403 error."""
```

File: synapse/handlers/message.py [hunk 4]
```diff
@@ -319,8 +319,10 @@ async def get_joined_members(self, requester: Requester, room_id: str) -> dict:
                 room_id, user_id, allow_departed_users=True
             )
             if membership != Membership.JOIN:
-                raise NotImplementedError(
-                    "Getting joined members after leaving is not implemented"
+                raise SynapseError(
+                    code=403,
+                    errcode=Codes.FORBIDDEN,
+                    msg="Getting joined members after leaving is forbidden.",
```
Comment (babolivier) [inline synapse/handlers/message.py:325]: The fact that the membership isn't `Membership.JOIN` doesn't necessarily mean the user has left the room (implying they have first joined it at some point). It just means you're not currently in the room (e.g. maybe you've never joined it, or you've been kicked, etc). This is kind of on the previous error message for getting it wrong though.
I would write something like "You must be a current member of the room to request its joined members." or "Getting joined members while not being a current member of the room is forbidden."

**Review by babolivier — CHANGES_REQUESTED** [review]
Thanks for this contribution! I've noted a few things that I think need changing but overall it looks pretty good to me!

File: changelog.d/13374.bugfix (hunk 1 above)
Comment (DMRobertson) [reply changelog.d/13374.bugfix:1]:
```suggestion
Fix a bug introduced in Synapse 0.24.0 where responses to `/joined_members` requests would have the wrong HTTP status code when the requester is not a current member of the room. Contributed by @andrewdoh.
```

**Review by babolivier — APPROVED** [review]
Sorry about the delay - thanks for your contribution!

**Comment by DMRobertson** [comment]
The PR title is

> Return 404 or member list when getting joined_members after leaving

But looking at the source this seems to only ever return 403 in this situation?

**Pull request merged**